### PR TITLE
Issue 3113: Kubernetes based system test framework to deploy and test Pravega.

### PR DESCRIPTION
**Pull request by shrids**
(Note: this PR should be merged only after PR: #3114  has been merged to master).
**Change log description**  
Ensure the Pravega tests can run on a K8 Cluster.  

**Purpose of the change**  
Fixes the following aspects of #3113 
- enable Pravega deployment using pravega operator and ensure related system tests are also run on the K8 cluster.
- enable `Service#scaleService( newinstanceCount)` api for bookkeeper, pravega-controller and pravega segment store.

**What the code does**  
The code ensures pravega related system tests can be run on a K8 cluster. 

**How to verify it**  
Pravega tests should run in a K8 Environment.

```
gradle --info startK8SystemTests -PpravegaVersion=<pravegaVersion> 
-PrepoUrl=<repo where the build artifacts are present> --tests io.pravega.test.system.ReaderCheckpointTest
```
Note: Validation is in progress.

### CODE REVIEW

**Comment by adrianmo** [comment]
@shrids I've implemented the ability to scale Pravega Controller in https://github.com/pravega/pravega-operator/pull/99

File: test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java [hunk 1]
```diff
@@ -22,21 +25,28 @@
     public enum TestExecutorType {
         LOCAL,
         DOCKER,
+        K8s,
```
Comment (adrianmo) [inline test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java]: For consistency and clarity, please use capital letters and the full name: `K8s` --> `KUBERNETES`.

File: test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java [hunk 2]
```diff
@@ -22,21 +25,28 @@
...
 
-    public TestExecutor getTestExecutor(TestExecutorType type) {
+    TestExecutor getTestExecutor(TestExecutorType type) {
         switch (type) {
             case DOCKER:
                 return getDockerExecutor();
             case REMOTE_SEQUENTIAL:
                 return getMarathonSequentialExecutor();
+            case K8s:
+                 return new K8SequentialExecutor();
```
Comment (adrianmo) [inline test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java]: Is there any reason why this executor is not obtained through a Getter like the Docker and Marathon ones?

E.g.:
```
    @Getter(lazy = true)
    private final TestExecutor kubernetesSequentialExecutor = new KubernetesBasedTestExecutor();
...
            case KUBERNETES:
                 return getKubernetesSequentialExecutor();
```

File: test/system/src/main/java/io/pravega/test/system/framework/Utils.java [hunk 3]
```diff
@@ -46,20 +55,42 @@ public static String getConfig(final String key, final String defaultValue) {
...
-        return DOCKER_BASED ? new ZookeeperDockerService("zookeeper")
-                : new ZookeeperService("zookeeper");
+        switch (EXECUTOR_TYPE) {
+            case REMOTE_SEQUENTIAL:
+                return new ZookeeperService("zookeeper");
+            case DOCKER:
+                return new ZookeeperDockerService("zookeeper");
+            case K8s:
+            default:
+                return new ZookeeperK8sService();
```
Comment (adrianmo) [inline test/system/src/main/java/io/pravega/test/system/framework/Utils.java]: For consistency, can you add the service name as a parameter?

File: test/system/src/main/java/io/pravega/test/system/framework/Utils.java [hunk 4]
```diff
@@ -46,20 +55,42 @@ public static String getConfig(final String key, final String defaultValue) {
...
 
     public static Service createZookeeperService() {
-        return DOCKER_BASED ? new ZookeeperDockerService("zookeeper")
-                : new ZookeeperService("zookeeper");
+        switch (EXECUTOR_TYPE) {
+            case REMOTE_SEQUENTIAL:
+                return new ZookeeperService("zookeeper");
+            case DOCKER:
+                return new ZookeeperDockerService("zookeeper");
+            case K8s:
```
Comment (adrianmo) [inline test/system/src/main/java/io/pravega/test/system/framework/Utils.java]: I would always use the full name instead of the abbreviation.

**Comment by adrianmo** [comment]
@shrids what `repoUrl` can I use to test the changes? I need it to be publicly available.

File: test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java (hunk 2 above)
Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/TestExecutorFactory.java]: updated the code to follow the same pattern.

File: test/system/src/main/java/io/pravega/test/system/framework/Utils.java (hunk 4 above)
Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/Utils.java]: done.

**Comment by shrids** [comment]
`repoUrl` can be any maven repository url.  `uploadArtifacts`  gradle task (https://github.com/pravega/pravega/blob/master/gradle/maven.gradle) can be used to upload the test artifact. This artifact is then downloaded on to a test pod where the specified test is executed.

**Pull request merged**